daemon/monitor.go

```diff
@@ -85,16 +85,29 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 		}
 		return daemon.postRunProcessing(c, e)
 	case libcontainerd.StateExitProcess:
+		closeAsync := make(chan bool)
+		waitClose := false
 		c.Lock()
-		defer c.Unlock()
-		if execConfig := c.ExecCommands.Get(e.ProcessID); execConfig != nil {
-			ec := int(e.ExitCode)
-			execConfig.ExitCode = &ec
-			execConfig.Running = false
-			execConfig.Wait()
-			if err := execConfig.CloseStreams(); err != nil {
-				logrus.Errorf("%s: %s", c.ID, err)
+		defer func() {
+			c.Unlock()
+			if waitClose {
+				<-closeAsync
```

Comment:
Actually, this synchronization should be removed altogether.

The routine would still wait until the exec IO is closed, which would prevent new events to be processed (at least on linux).

WDYT @tonistiigi ?